// zetapush-js/es/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _handshake = require('./authentication/handshake');

Object.defineProperty(exports, 'Authentication', {
  enumerable: true,
  get: function () {
    return _handshake.Authentication;
  }
});

var _connectionStatus = require('./connection/connection-status');

Object.defineProperty(exports, 'ConnectionStatusListener', {
  enumerable: true,
  get: function () {
    return _connectionStatus.ConnectionStatusListener;
  }
});

var _basic = require('./client/basic');

Object.defineProperty(exports, 'Client', {
  enumerable: true,
  get: function () {
    return _basic.Client;
  }
});

var _smart = require('./client/smart');

Object.defineProperty(exports, 'SmartClient', {
  enumerable: true,
  get: function () {
    return _smart.SmartClient;
  }
});

var _weak = require('./client/weak');

Object.defineProperty(exports, 'WeakClient', {
  enumerable: true,
  get: function () {
    return _weak.WeakClient;
  }
});

var _server = require('./client/server');

Object.defineProperty(exports, 'ServerClient', {
  enumerable: true,
  get: function () {
    return _server.ServerClient;
  }
});

var _mapping = require('./mapping');

Object.defineProperty(exports, 'services', {
  enumerable: true,
  get: function () {
    return _mapping.services;
  }
});


/**
 * SDK Version
 * @type {string}
 */
const VERSION = exports.VERSION = '3.4.0-alpha.2';
